### DISCUSSION TITLE
_CHAR_256 and _char_256 impl different

### DISCUSSION THREAD

**Discussion by Lightczx** [microsoft/CsWin32#891] [Ideas]
`_CHAR_256` use raw `CHAR` fields, like `_0,_1,_2,_3,_4,_5,_6,_7,_8,_9`
while `_char_256` use a fixed char array.
Should we unify this?

**Comment by AArnott** [comment]
No, these cannot be unified. `CHAR` is an 1-byte value whereas `char` is a UTF-16 value.

**Reply by Lightczx** [reply]
So why can't use array in _CHAR_256 then?
I'm still confused.

**Reply by AArnott** [reply]
These structures are not .NET arrays, which would be allocated on the heap. These structs store a 'fixed array' which means the content of the 'array' is actually _inline_ in the struct itself. These fixed array structs do include some helper members in them (which vary based on your TargetFramework and C# language version) to help make them convenient to read and write though.

**Reply by Lightczx** [reply]
Maybe I didn't express it clearly.
I do know CHAR is one byte wide and char is two bytes wide.

I mean, can we use `public unsafe fixed winmdroot.Foundation.CHAR[] Value[SpanLength]` like in _char_256?

**Reply by AArnott** [reply]
No, because [C# fixed arrays only allow a small set of primitive types](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#fixed-size-buffers).

**Reply by Lightczx** [reply]
I see, thanks for your explanation.
